Objective Function Refactor

ashelkovnykov opened the pull request:
PR for issue #122 

Major changes include:
- Split `LBFGS` and `OWLQN` optimizers
- Removed all `withL1Regularization()` / `withL2Regularization()` code from the differentiable traits
- Removed the `L1RegularizationTerm` trait (choosing between `LBFGS` and `OWLQN` happens in the `OptimizerFactory` using the `RegularizationContext`
- Moved L2 regularization code to the stackable `L2Regularization` trait
- Created the `ObjectiveFunction` class as a base for the `DiffFunction`, `TwiceDiffFunction`, and `L2Regularization` traits
- RDD ownership has been removed wherever possible, working towards a pipeline of transformers
- The optimization problems and objective functions have been split into one class for Iterable[LabeledPoint] and one class for RDD[LabeledPoint]
- The `AbstractOptimizer` and `Optimizer` have been merged into a single class
- The hacky typing and contravariance issues from the first refactor have been resolved

**TODO:** Performance tests between Photon and GAME from before the refactor

joshvfleming commented on photon-ml/src/test/scala/com/linkedin/photon/ml/optimization/LinearRegressionOptimizationProblemTest.scala:
```diff
@@ -63,15 +63,15 @@ class LinearRegressionOptimizationProblemTest {
 
 object LinearRegressionOptimizationProblemTest {
-  val Dimensions = 10
+  val DIMENSIONS = 10
```
I know it's inconsistent with how we've done things here, but the Scala style guide actually recommends upper camel case for constants:

[http://docs.scala-lang.org/style/naming-conventions.html#constants-values-variable-and-methods](http://docs.scala-lang.org/style/naming-conventions.html#constants-values-variable-and-methods)

I've been trying to nudge us in that direction, but we should probably make a wholesale change, or just have them all use the Java convention.

joshvfleming commented on photon-ml/src/test/scala/com/linkedin/photon/ml/data/LabeledPointTest.scala:
```diff
@@ -89,10 +96,56 @@ class LabeledPointTest {
+  {
+    // Init random coefficients
+    val r: Random = new Random(LabeledPointTest.PARAMETER_RANDOM_SEED)
+    val coef: Vector[Double] = DenseVector.fill[Double](DataPointTest.PROBLEM_DIMENSION){ r.nextDouble() }
```
There's a utility function for this:

https://github.com/linkedin/photon-ml/blob/master/photon-test/src/main/scala/com/linkedin/photon/ml/test/CommonTestUtils.scala#L75

joshvfleming commented on photon-ml/src/integTest/scala/com/linkedin/photon/ml/algorithm/RandomEffectCoordinateTest.scala:
```diff
@@ -55,17 +58,17 @@ class RandomEffectCoordinateTest extends SparkTestUtils with GameTestUtils {
-  val NumPartitions = 4
+  val NUM_TRAINING_SAMPLES = 1000
+  val DIMENSIONALITY = 10
+  val NUM_PARTITIONS = 4
```
I know it's inconsistent with how we've done things here, but the Scala style guide actually recommends upper camel case for constants:

http://docs.scala-lang.org/style/naming-conventions.html#constants-values-variable-and-methods

I've been trying to nudge us in that direction, but we should probably make a wholesale change, or just have them all use the Java convention.

ashelkovnykov replied:
I wasn't aware. We should stick to the Scala style guide. I'll definitely change these back (as well as any other such changes in the code) just so we don't forget, in case we end up doing it piecemeal (as with the author tags).

XianXing replied:
Regarding the code style issue, actually I guess it might be a good idea to follow [Spark's Scala code style for](https://github.com/databricks/scala-style-guide), instead of the Scala style guide, if we are still considering to contribute some portion of Photon ML back to Spark.

However, please note that Spark's guide on Scala code style is closer to the Java convention and somewhat different from the Scala style guide, for example, [Spark prefers the constants to be in uppercases](https://github.com/databricks/scala-style-guide#syntactic-style).

There are a few noticeable differences (and I should've mentioned them earlier), for example, the [doc style](https://github.com/databricks/scala-style-guide#doc).

Actually this is why you may have noticed that the code style of Photon ML sometimes doesn't follow the Scala code style guide strictly.

joshvfleming replied:
Ah, good point @XianXing. Let's stick with the uppercase convention for constants. And in general, we should probably follow Spark style guidelines when they conflict with Scala style.

ashelkovnykov replied:
Excellent link @XianXing. I wish I saw that earlier, as I've been undoing some Spark style in situations where it conflicted with Scala style.

@joshvfleming +1

ashelkovnykov commented on photon-ml/src/main/scala/com/linkedin/photon/ml/ModelTraining.scala:
```diff
@@ -155,42 +160,34 @@ object ModelTraining extends Logging {
           // Initialize the list with the result from the first regularization weight
-          val updatedOptimizationProblem = optimizationProblem.updateObjective(
-            normalizationContextProvider, currentWeight)
+          optimizationProblem.updateRegularizationWeight(currentWeight)
```
@joshvfleming 
Updating regularization weight is stateful now. I did it because it's the easy way out, though I would not object to going back through the code and removing state wherever possible.

XianXing replied:
Thanks for highlighting the change. Does it mean the optimization problem can be reused across different runs of model training with different regularization weights?

ashelkovnykov replied:
Yes, the update the regularization weight for the problem and re-run. It was easier to add in regularization weight updates as a change to state, and in line with optimization state tracking which is also stateful. However, as mentioned above, I'm not opposed to going through and making the design functional.

XianXing replied:
No problem, I am just assisting Josh reviewing the PR and want to learn more context, and you guys can make the final call :)

ashelkovnykov commented on photon-ml/src/main/scala/com/linkedin/photon/ml/algorithm/FactoredRandomEffectCoordinate.scala:
```diff
@@ -28,22 +27,29 @@ import com.linkedin.photon.ml.util.VectorUtils
 
 /**
-  * The optimization problem coordinate for a factored random effect model
+  * The optimization problem coordinate for a random effect model using matrix factorization
```
@joshvfleming 
The Coordinate files look like they've been substantially changed on GitHub, but mostly it was moving the order of functions around to be consistent between classes. This may be easier to see by checking out the PR locally.

ashelkovnykov commented on photon-ml/src/main/scala/com/linkedin/photon/ml/function/DiffFunction.scala:
```diff
@@ -15,307 +15,38 @@
+    * @param coefficients The model coefficients used to compute the function's value and gradient
+    * @return The computed value and gradient of the function
+    */
+  protected[ml] def calculate(coefficients: Vector[Double]): (Double, Vector[Double])
```
@joshvfleming 
DiffFunction is just an interface now, all of the bulk has been stripped out (same goes for TwiceDiffFunction).

ashelkovnykov commented on photon-ml/src/main/scala/com/linkedin/photon/ml/function/L1RegularizationTerm.scala:
```diff
@@ -1,22 +0,0 @@
- */
-trait L1RegularizationTerm {
-  def getL1RegularizationParam: Double
-}
```
@joshvfleming 
L1Regularization trait is gone now; treated as a special case when selecting an optimizer.

ashelkovnykov commented on photon-ml/src/main/scala/com/linkedin/photon/ml/function/L2Regularization.scala:
```diff
@@ -0,0 +1,106 @@
+    * @return The Hessian diagonal of the L2 regularization term
+    */
+  protected def l2RegHessianDiagonal: Double = l2RegWeight
+}
```
@joshvfleming 
Moved the old 'withL2Regularization' code into these stackable traits. I'm not sure if this is correct style or not, but it seemed so because unfortunately, as we discussed offline, each descendant of the ObjectiveFunction class needs its own unique behaviour (since each one introduces new methods).

PS - woops, no ending newline

joshvfleming replied:
Nice!

ashelkovnykov commented on photon-ml/src/main/scala/com/linkedin/photon/ml/function/ObjectiveFunction.scala:
```diff
@@ -0,0 +1,54 @@
+  *                           aggregation. It is experimental currently because treeAggregate is unstable in Spark
+  *                           versions 1.4 and 1.5.
+  */
+abstract class ObjectiveFunction(
```
@joshvfleming 
The new base class for objective functions. Biggest change is that the function is now the owner of the data that it scores.

XianXing replied:
What is the motivation/benefit of letting the objective function own the data?

Including ```RDD``` typed values/variables within a class is tricky, as it requires careful considerations on the serialization/deserialization of all values/variables within that class.

joshvfleming replied:
Yeah, I second @XianXing's concern. It's also little odd structurally to have a stateful function.

ashelkovnykov replied:
I spoke with @XianXing offline about the potential performance implications of this change, and the differences between moving the process to the data and moving the data to the process. 

I'm interested to see what the performance hit of this change will be during testing, and I'm prepared to revert it.

ashelkovnykov commented on photon-ml/src/main/scala/com/linkedin/photon/ml/function/SmoothedHingeLossFunction.scala:
```diff
@@ -14,19 +14,83 @@
+  *                           aggregation. It is experimental currently because treeAggregate is unstable in Spark
+  *                           versions 1.4 and 1.5.
+  */
+class SmoothedHingeLossFunction(
```
@joshvfleming 
The default behaviour code that was in DiffFunction was only used by the SmoothedHingeLossFunction and overwritten by the GLMLossFunction, so I moved it into the SmoothedHingeLossFunction and as previously mentioned turned DiffFunction into a pure interface.

joshvfleming replied:
I think we'll want to keep this implementation in DiffFunction. The reason: as of right now, the only implementation is SmoothedHingeLossFunction, but we'd like to support easily adding new loss functions. With this change, if someone adds a new loss function that happens to be only first-order continuous, they'll need to extend SmoothHingeLossFunction.

ashelkovnykov commented on photon-ml/src/main/scala/com/linkedin/photon/ml/optimization/AbstractOptimizer.scala:
```diff
@@ -1,132 +0,0 @@
-      case None =>
-    }
-  }
-}
```
@joshvfleming 
AbstractOptimizer has been nuked. The distinction between it and Optimizer seemed arbitrary. I'd like your opinion on this.

XianXing replied:
When Brendan designed this class, the motivation was to let ```AbstractOptimizer``` to handle the bookkeeping of the optimizer's inner states, because such bookkeeping was independent of the optimizer's core logic: optimization. And the plan was that in the future if we find a better way to handle the inner states, only components in ```AbstractOptimizer``` needs to be changed with the optimizer remaining intact.

ashelkovnykov replied:
My concern is that the split between the two (the Optimizer trait and the AbstractOptimizer class) wasn't actually accomplishing this. While in theory an optimizer's scope should be limited to optimization, the interface defined by the Optimizer trait defined methods such as 'setInitialState', 'setCurrentState', and 'setPreviousState' which suggest that tracking state is integral to the Optimizer.

Thus the only reason to not implement the state tracking methods within Optimizer, to push that out to AbstractOptimizer, is if we're going to have multiple ways for managing internal states. And even then, since this is behaviour that's tangential to optimization, the optimizer implementations (LBFGS, OWLQN, TRON) should not descend from a specific state tracking system: it should be a stackable trait that overrides a default state tracking behaviour, whatever that may be.

ashelkovnykov commented on photon-ml/src/main/scala/com/linkedin/photon/ml/optimization/LBFGS.scala:
```diff
@@ -15,126 +15,99 @@
-      case diffFunc: DiffFunction[Datum] =>
-        new BreezeLBFGS[Vector[Double]](maxNumIterations, numCorrections, tolerance)
-    }
+  protected val breezeOptimizer = new BreezeLBFGS[Vector[Double]](maxNumIterations, numCorrections, tolerance)
```
@joshvfleming 
OWLQN has become its own optimizer. The selection between LBFGS and OWLQN is now made within the OptimizerFactory.

ashelkovnykov commented on photon-ml/src/main/scala/com/linkedin/photon/ml/optimization/Optimizer.scala:
```diff
@@ -14,252 +14,218 @@
+  val DEFAULT_CONSTRAINT_MAP = None
+  val DEFAULT_TRACKING_STATE = true
+  val DEFAULT_REUSE_PREVIOUS_INIT_STATE = true
+}
```
@joshvfleming 
Both the non-Scala-style constant names and missing newline are noted.

ashelkovnykov commented on photon-ml/src/main/scala/com/linkedin/photon/ml/optimization/OptimizerFactory.scala:
```diff
@@ -14,59 +14,68 @@
-    val optimizer = config.optimizerType match {
-      case OptimizerType.LBFGS =>
-        new LBFGS[LabeledPoint]
+  def combinedOptimizer(
```
@joshvfleming 
As discussed, single function now for generation of optimizers. Optimizer and objective function are now generated independently and should align at runtime.

joshvfleming replied:
LGTM, bit let's call this "createOptimizer".

ashelkovnykov commented on photon-ml/src/test/scala/com/linkedin/photon/ml/function/ObjectiveFunctionTest.scala:
```diff
@@ -16,343 +16,501 @@ package com.linkedin.photon.ml.function
-          )
-        case _ => Seq()
-      }
+  def getDifferentiableFunctions: Array[Array[DiffFunction]] = diffTasks.flatMap {
```
@joshvfleming 
Generate test data in a more programmatic way.

ashelkovnykov commented on photon-ml/src/test/scala/com/linkedin/photon/ml/function/ObjectiveFunctionTest.scala:
```diff
@@ -16,343 +16,501 @@ package com.linkedin.photon.ml.function
+    * @param expected The computed value of the derivative
+    * @param tolerance The relative & absolute error tolerance
+    */
+  def checkDerivativeError(
```
@joshvfleming 
Fused 'checkGradient' and 'checkHessian' functions into one. They were both estimating the derivative in the same way.

ashelkovnykov commented on photon-ml/src/test/scala/com/linkedin/photon/ml/optimization/LBFGSTest.scala:
```diff
@@ -16,115 +16,66 @@ package com.linkedin.photon.ml.optimization
+      Array(None, Array(4.0, 1.0)),
 
-      // note that expected value here is the value of the function at the unconstrained optima since the
+      // Note that expected value here is the value of the function at the unconstrained optima since the
```
@joshvfleming 
Removed regularization weight (that's tested for OWLQN now) and expected value (it should be static now that the regularization weight is fixed at 0). Played around with the test values to be clearer that we're checking that they're being bounded correctly.

ashelkovnykov commented on photon-ml/src/test/scala/com/linkedin/photon/ml/optimization/OWLQNTest.scala:
```diff
@@ -0,0 +1,85 @@
+  * (x-1)^2 + (y-4)^2 + C(x+y) which has obvious analytic solution. The test is based on this function and verifies the
+  * shrinkage of x and y.
+  */
+class OWLQNTest {
```
@joshvfleming 
More-or-less a copy of LBFGSTest with the other half of the tests.

ashelkovnykov commented on photon-ml/src/test/scala/com/linkedin/photon/ml/optimization/TestObjective.scala:
```diff
@@ -14,44 +14,78 @@
-class TestObjective extends TwiceDiffFunction[LabeledPoint] {
+  * Test objective function used solely to exercise the optimizers.
+  */
+abstract class TestObjective(myData: Either[RDD[LabeledPoint], Iterable[LabeledPoint]])
```
@joshvfleming 
This class has been heavily modified for the new objective function class stack, though it's the same in spirit (a toy objective function with an optimal value at known coordinates). It's used by the LBFGS and OWLQN tests now, instead of the ObjectiveFunctionTest.

ashelkovnykov commented on photon-ml/src/test/scala/com/linkedin/photon/ml/function/ObjectiveFunctionTest.scala:
```diff
@@ -16,343 +16,501 @@ package com.linkedin.photon.ml.function
- */
+  * Unit tests to verify that the loss functions compute gradients & Hessians accurately.
+  */
 class ObjectiveFunctionTest extends SparkTestUtils {
```
@joshvfleming 
Needs to be renamed, now that there is an actual ObjectiveFunction class. TBH, this class is more of an integration test than a unit test.

ashelkovnykov commented on photon-ml/src/integTest/scala/com/linkedin/photon/ml/function/ObjectiveFunctionIntegTest.scala:
```diff
@@ -16,402 +16,554 @@ package com.linkedin.photon.ml.function
- */
+  * Integration tests to verify that the loss functions compute gradients & Hessians accurately.
+  */
 class ObjectiveFunctionIntegTest extends SparkTestUtils {
```
@joshvfleming 
Very similar changes to that of the ObjectiveFunctionTest unit tests class.

XianXing commented on photon-ml/src/main/scala/com/linkedin/photon/ml/function/DiffFunction.scala:
```diff
@@ -15,307 +15,38 @@
+  *
+  * FAQ: Why do the functions take an ObjectProvider[ Vector[Double] ] for coefficients instead of a Vector[Double]?
+  *      Why may the functions potentially receive a Broadcast[ Vector[Double] ]?
+  * A: In the optimization procedure, the gradient/value/calculate function might be called repeatedly several times
```
The FAQ looks confusing to me, please update it.
1. Which functions take ObjectProvider[ Vector[Double] ]?
2. Where is Broadcast[ Vector[Double] ] being used?

XianXing commented on photon-ml/src/main/scala/com/linkedin/photon/ml/function/ObjectiveFunction.scala:
```diff
@@ -0,0 +1,54 @@
+import org.apache.spark.rdd.RDD
+
+/**
+  * The objective loss function for an optimization problem.
```
objective loss function -> objective function

XianXing commented on photon-ml/src/main/scala/com/linkedin/photon/ml/function/ObjectiveFunction.scala:
```diff
@@ -0,0 +1,54 @@
+  *                           versions 1.4 and 1.5.
+  */
+abstract class ObjectiveFunction(
+    protected[ml] val data: Either[RDD[LabeledPoint], Iterable[LabeledPoint]],
```
We (Brendan, Yitong and I) once discussed a refactoring plan on ```DiffFunction``` to get rid of ```Either``` and use polymorphism instead in such scenario.

I should've discussed this with you earlier, because this ```DiffFunction``` redesign and refactoring was a good opportunity to make it happen. But let's leave ```Either``` as is in this PR and address this problem later.

XianXing commented on photon-ml/src/main/scala/com/linkedin/photon/ml/function/GeneralizedLinearModelLossFunction.scala:
```diff
@@ -32,137 +32,109 @@ import org.apache.spark.rdd.RDD
+    case Left(dataAsRDD) =>
+      val broadcastCoefficients = dataAsRDD.sparkContext.broadcast(coefficients)
+      val broadcastMultiply = dataAsRDD.sparkContext.broadcast(multiplyVector)
+      HessianVectorAggregator.calcHessianVector(
```
Please read the FAQ of the original ```DiffFunction``` on whether we want to re-broadcast everything or not.

Or, please let me known the motivation of making such changes.

XianXing commented:
This PR involves more changes than described in the [design proposal](https://docs.google.com/document/d/1g718CGM1W6aKlfb79pyuHUK7s9GYtrVawOPAWEpnwGM/edit), and many of the changes are performance critical.

Let's please benchmarking the performance of Photon-ML before and after this PR before merging it to the repo.

joshvfleming commented on photon-ml/src/main/scala/com/linkedin/photon/ml/algorithm/RandomEffectCoordinateInProjectedSpace.scala:
```diff
@@ -26,29 +26,40 @@ import com.linkedin.photon.ml.supervised.model.GeneralizedLinearModel
-    */
-  protected[algorithm] override def updateModel(model: DatumScoringModel): (DatumScoringModel, OptimizationTracker) =
+  override protected[algorithm] def updateCoordinateWithDataSet(updatedRandomEffectDataSet: RandomEffectDataSet)
+    : RandomEffectCoordinate[GLM, F] = {
```
The style isn't quite right here. Newline should start with argument name.

joshvfleming commented on photon-ml/src/main/scala/com/linkedin/photon/ml/optimization/LogisticRegressionOptimizationProblem.scala:
```diff
@@ -14,179 +14,183 @@
-      variances: Option[Vector[Double]]): LogisticRegressionModel =
-    new LogisticRegressionModel(Coefficients(coefficients, variances))
+  override protected[optimization] def createModel(coefficients: Vector[Double], variances: Option[Vector[Double]])
+    : LogisticRegressionModel = new LogisticRegressionModel(Coefficients(coefficients, variances))
```
Scala style, newline with params.

ashelkovnykov commented:
This PR has been majorly updated as of now. Please take a look through it once again. I've broken it into 4 commits, but if it's still too daunting to do easily please let me know and I will go through and mark the largest changes (as I did with the first draft).

This took a long time and by the end I went with the most convenient solution, maybe not the best! Be vigilant!

ashelkovnykov commented on photon-ml/src/main/scala/com/linkedin/photon/ml/function/IndividualObjectiveFunction.scala:
```diff
@@ -0,0 +1,55 @@
+ * The base objective function used by IndividualOptimizationProblems. This function works with data locally as part of
+ * a single task (on a single executor).
+ */
+abstract class IndividualObjectiveFunction extends ObjectiveFunction with Serializable {
```
See 'DistributedObjectiveFunction'.

XianXing replied:
I like the overall design. Minor note (related to #60): ```IndividualObjectiveFunction``` sounds a little bit odd to me, why not call it something like ```LocalObjectiveFunction``` or ```SingleObjectiveFunction```?

TDDFT replied:
+1 for @XianXing 's comments. It sounds really odd to me. The opposite of `DistributedObjectiveFunction` is `SingleMachineObjectiveFunction`. It doesn't sound very good but it kinds of reflects what it is.

joshvfleming replied:
+1

Another possibility: ```SingleNodeObjectiveFunction```

Sigh. Naming things is hard.

ashelkovnykov replied:
I thought of `LocalObjectiveFunction` but Josh pointed out that the overall problem is still distributed, it's just only used on one machine.

I like `SingleObjectiveFunction` or `SingleNodeObjectiveFunction`. Please continue to voice your preferences, otherwise I'll just pick one of those two.

XianXing replied:
Let do ```IndividualObjectiveFunction```.

XianXing replied:
By saying ```IndividualObjectiveFunction ``` above I actually meant ```SingleNodeObjectiveFunction```. I recall that Alex checked this with me and I verbally corrected myself, but anyway, just want to make it clear here.

ashelkovnykov commented on photon-ml/src/main/scala/com/linkedin/photon/ml/function/L2Regularization.scala:
```diff
@@ -0,0 +1,175 @@
+/**
+ * Trait for an objective function with L2 regularization.
+ */
+trait L2Regularization extends ObjectiveFunction {
```
Old L2 regularization code has been moved into these stackable traits.

ashelkovnykov commented on photon-ml/src/main/scala/com/linkedin/photon/ml/function/ObjectiveFunction.scala:
```diff
@@ -0,0 +1,73 @@
+/**
+ * The base objective function class for an optimization problem.
+ */
+abstract class ObjectiveFunction {
```
The new base class for representing objective functions in an optimization problem. All of the additional mixin traits (DiffFunction, TwiceDiffFunction, L2Regularization) extend this class.

ashelkovnykov commented on photon-ml/src/main/scala/com/linkedin/photon/ml/function/ObjectiveFunction.scala:
```diff
@@ -0,0 +1,73 @@
+   * @param coefficients A coefficients Vector to convert
+   * @return The given coefficients Vector as an object of the type expected by the function
+   */
+  protected[ml] def convertFromVector(coefficients: Vector[Double]): Coefficients
```
There are two major variants of ObjectiveFunction currently: IndividualObjectiveFunction, used by random effect problems to solve an optimization problem local to one executor, and DistributedObjectiveFunction, used by fixed effect problems to solve an optimization problem across several tasks on one or more executors.

The optimizers, however, are oblivious to what kind of problem they are solving. They use these 3 functions (convertFromVector, convertToVector, and cleanupCoefficients) to be agnostic. For IndividualObjectiveFunctions, these functions do nothing. For DistributedObjectiveFunctions, they broadcast coefficients and unpersist them afterwards.

ashelkovnykov commented on photon-ml/src/main/scala/com/linkedin/photon/ml/function/TwiceDiffFunction.scala:
```diff
@@ -15,389 +15,36 @@
+/**
+ * Trait for twice differentiable function.
+ */
+trait TwiceDiffFunction extends DiffFunction {
```
See DiffFunction.

ashelkovnykov commented on photon-ml/src/main/scala/com/linkedin/photon/ml/function/glm/DistributedGLMLossFunction.scala:
```diff
@@ -0,0 +1,168 @@
+ *                           aggregation. It is experimental currently because treeAggregate is unstable in Spark
+ *                           versions 1.4 and 1.5.
+ */
+protected[ml] class DistributedGLMLossFunction private (
```
Old GeneralizedLinearLossFunction class split into two variants (just like the optimization problems and objective functions).

ashelkovnykov commented on photon-ml/src/main/scala/com/linkedin/photon/ml/function/glm/IndividualGLMLossFunction.scala:
```diff
@@ -0,0 +1,151 @@
+ *
+ * @param singleLossFunction A single loss function l(z, y) used for the generalized linear model
+ */
+protected[ml] class IndividualGLMLossFunction private (singleLossFunction: PointwiseLossFunction)
```
See DistributedGLMLossFunction.

ashelkovnykov commented on photon-ml/src/main/scala/com/linkedin/photon/ml/function/svm/DistributedSmoothedHingeLossFunction.scala:
```diff
@@ -0,0 +1,132 @@
+ *                           aggregation. It is experimental currently because treeAggregate is unstable in Spark
+ *                           versions 1.4 and 1.5.
+ */
+protected[ml] class DistributedSmoothedHingeLossFunction(sparkContext: SparkContext, treeAggregateDepth: Int)
```
See the DistributedGLMLossFunction.

ashelkovnykov commented on photon-ml/src/main/scala/com/linkedin/photon/ml/function/svm/IndividualSmoothedHingeLossFunction.scala:
```diff
@@ -0,0 +1,112 @@
+ * A:   Using cumGradient allows the functions to avoid memory allocation by modifying and returning cumGradient instead
+ *      of creating a new gradient vector.
+ */
+protected[ml] class IndividualSmoothedHingeLossFunction extends IndividualObjectiveFunction with DiffFunction {
```
See the IndividualGLMLossFunction.

ferd36 replied:
This "Individual" - I don't find very clear. I would prefer "SingleHost" or "Local". "Local" might be the best, people are familiar with "local/distributed". "Individual" to me is connected to "one" like "individual/group", so it's not terribly appropriate here, I think.

ashelkovnykov commented on photon-ml/src/main/scala/com/linkedin/photon/ml/optimization/DistributedOptimizationProblem.scala:
```diff
@@ -0,0 +1,198 @@
+ * @param regularizationContext The regularization context
+ * @param isComputingVariances Should coefficient variances be computed in addition to the means?
+ */
+protected[ml] class DistributedOptimizationProblem[Function <: DistributedObjectiveFunction] protected[optimization] (
```
Split the optimization problems into two variants: IndividualOptimizationProblem and DistributedOptimizationProblem.

IndividualOptimizationProblems are used to solve the individual random effect optimization problems local to one task.

DistributedOptimizationProblems are used to solve fixed effect optimization problems across several tasks on one or more executors. They can also downsample input data prior to each run.

This split helps to enforce Photon/GAME as a pipeline of transformers. It also makes it easier to quickly determine which classes should be serializable and which should not.

XianXing replied:
Nice design choice.

ashelkovnykov commented on photon-ml/src/main/scala/com/linkedin/photon/ml/optimization/IndividualOptimizationProblem.scala:
```diff
@@ -0,0 +1,139 @@
+ * @param glmConstructor The function to use for producing GLMs from trained coefficients
+ * @param isComputingVariances Should coefficient variances be computed in addition to the means?
+ */
+protected[ml] class IndividualOptimizationProblem[Function <: IndividualObjectiveFunction] protected[optimization] (
```
See DistributedOptimizationProblem.

ashelkovnykov commented on photon-ml/src/main/scala/com/linkedin/photon/ml/optimization/OWLQN.scala:
```diff
@@ -0,0 +1,91 @@
+ *                                      desired, i.e. in grid-search based hyper-parameter tuning, this field is
+ *                                      recommended to set to true for consistent convergence check.
+ */
+class OWLQN(
```
OWLQN is its own optimizer now. The OptimizerFactory determines which one to use at runtime based on the RegularizationContext provided by the user.

ashelkovnykov commented on photon-ml/src/main/scala/com/linkedin/photon/ml/data/ObjectProvider.scala:
```diff
@@ -1,42 +0,0 @@
-@SerialVersionUID(1L)
-class BroadcastedObjectProvider[T <: Serializable](obj: Broadcast[T]) extends ObjectProvider[T] {
-  override def get: T = obj.value
-}
```
Now that there is an explicit split between optimization problems working with local datasets and optimization problems working with distributed datasets, the ObjectProvider class is no longer necessary.

ashelkovnykov replied:
The two code paths were originally overlapping because in the end the code they want to call is the same: it's the code in the `ValueAndGradientAggregator` and `HessianVectorAggregator`. For fixed effect problems, we need to perform a distributed aggregation. For random effect problems, it is a localized one.

However, to merge the logic paths in one class meant that the class either had to have multiple identical methods (one for each input data type) or it had to take input objects of multiple data types in a wrapper object. Currently it does the latter; some objects using the `ObjectProvider`, and some using `Either`.

In addition, the logic of the two problems is not completely identical. The fixed effect problem should **have** an RDD of data, should not be serializable, and should perform down-sampling. The individual random effect problems should **be** an RDD of optimization problems that handle data locally and be serializable.

It is my opinion that these differences are substantial enough (the serialization relationship alone) that they should be made explicit in the structure of the code.

ashelkovnykov commented on photon-ml/src/test/scala/com/linkedin/photon/ml/function/ObjectiveFunctionTest.scala:
```diff
@@ -1,375 +0,0 @@
-  val PARAMETER_RANDOM_SEED: Int = 500
-  val TRAINING_SAMPLES = PROBLEM_DIMENSION * PROBLEM_DIMENSION
-  val LOGGER: Logger = LogManager.getLogger(classOf[ObjectiveFunctionTest])
-}
```
This test is renamed to the IndividualObjectiveFunctionTest.

ashelkovnykov commented on photon-ml/src/test/scala/com/linkedin/photon/ml/optimization/GeneralizedLinearOptimizationProblemTest.scala:
```diff
@@ -14,100 +14,215 @@
+  import GeneralizedLinearOptimizationProblemTest._
 
   @Test
-  def testRun(): Unit = {
```
Moved to the IndividualOptimizationProblemTest and DistributedOptimizationProblemIntegTest.

ashelkovnykov commented on photon-ml/src/test/scala/com/linkedin/photon/ml/optimization/GeneralizedLinearOptimizationProblemTest.scala:
```diff
@@ -14,100 +14,215 @@
   }
 
   @Test
-  def testGetRegularizationTermValue: Unit = {
```
Moved to the DistributedOptimizationProblemIntegTest.

ashelkovnykov commented on photon-ml/src/test/scala/com/linkedin/photon/ml/optimization/LBFGSTest.scala:
```diff
@@ -16,115 +16,70 @@ package com.linkedin.photon.ml.optimization
     DenseVector(Seq.fill(dim)(random.nextGaussian).toArray)
   }
 
-  @Test(dataProvider = "dataProvider")
```
The L1 regularization code has been moved to the OWLQNTest.

ashelkovnykov commented on photon-ml/src/test/scala/com/linkedin/photon/ml/optimization/OptimizerTest.scala:
```diff
@@ -0,0 +1,192 @@
+/**
+ * Verify that core optimizers do reasonable things on small test problems.
+ */
+class OptimizerTest extends SparkTestUtils with Logging {
```
Moved the local tests here from the OptimizerIntegTest.

ashelkovnykov commented on photon-ml/src/integTest/scala/com/linkedin/photon/ml/function/ObjectiveFunctionIntegTest.scala:
```diff
@@ -1,431 +0,0 @@
- * @author bdrew
- * @author yali
- */
-class ObjectiveFunctionIntegTest extends SparkTestUtils {
```
Renamed to the DistributedOptimizationProblemIntegTest.

ashelkovnykov commented on photon-ml/src/integTest/scala/com/linkedin/photon/ml/optimization/OptimizerIntegTest/OptimizerIntegTest.scala:
```diff
@@ -1,195 +0,0 @@
-  * Verify that core optimizers do reasonable things on small test problems.
-  * @author bdrew
-  */
-class OptimizerIntegTest extends SparkTestUtils with Logging {
```
File moved - for some reason it was in its own folder.

XianXing commented on photon-ml/src/main/scala/com/linkedin/photon/ml/normalization/NormalizationContext.scala:
```diff
@@ -158,6 +158,8 @@ private[ml] object NormalizationContext {
-private[ml] object NoNormalization extends NormalizationContext(factors = None, shifts = None, interceptId = None)
+private[ml] object NoNormalization {
+  def apply(): NormalizationContext = NormalizationContext(factors = None, shifts = None, interceptId = None)
+}
```
@TDDFT Please take a look at the changes made on feature normalization when you have time.

TDDFT commented:
Start looking at it.

XianXing commented on photon-ml/src/main/scala/com/linkedin/photon/ml/ModelTraining.scala:
```diff
@@ -155,42 +163,36 @@ object ModelTraining extends Logging {
-          (updatedOptimizationProblem,
-            (currentWeight, updatedOptimizationProblem.run(trainingData, previousModel, normalizationContext))
-              :: latestWeightsAndModels)
+          (currentWeight, optimizationProblem.run(trainingData, previousModel)) +: latestWeightsAndModels
```
+@TDDFT

@ashelkovnykov Can you briefly describe the changes made with feature normalization context?

Since we have plans to support feature normalization/scaling in GAME soon, it would be great if you can help bring folks to the same page.

ashelkovnykov replied:
@TDDFT @XianXing 

There are no major changes to the normalization logic. The first change is that the NoNormalization object is no longer a singleton, derived from a NormalizationContext object set to no normalization. It is now a factory object for NormalizationContext objects set to no normalization. This change was made due to the second change: ObjectProvider has been removed, now that local and distributed optimization problems have been explicitly separated. The Spark Broadcast class is invariant, thus the old setup for NoNormalization did not work because Broadcast[NoNormalization] != Broadcast[NormalizationContext].

ashelkovnykov commented:
A suggestion for reviewing this PR:
Try going through the commits one at a time, while exploring the changes locally in IDEA. You can find instructions on how to pull a GitHub PR to your local branch [here](https://help.github.com/articles/checking-out-pull-requests-locally/) (the second section, "Modifying an inactive pull request locally", is the most useful, in spite of its inaccurate title). You'll want to modify the example command to look something like this:
`git fetch upstream pull/131/head:[branch name]`

XianXing commented on photon-ml/src/main/scala/com/linkedin/photon/ml/sampler/BinaryClassificationDownSampler.scala:
```diff
@@ -29,8 +30,7 @@ import scala.util.hashing.byteswap64
 protected[ml] class BinaryClassificationDownSampler(downSamplingRate: Double) extends DownSampler with Serializable {
-
-  require((downSamplingRate > 0D) && (downSamplingRate <= 1D), s"Invalid down-sampling rate: $downSamplingRate")
+  require((downSamplingRate > 0D) && (downSamplingRate < 1D), s"Invalid down-sampling rate: $downSamplingRate")
```
Minor issue, but why make this change? ```downSamplingRate``` can be 1D sometimes, and actually the default for ```downSamplingRate``` is ```1D```.

XianXing replied:
Nice, I didn't know that ```None``` is used when ```downSamplingRate``` is 1D.

XianXing commented on photon-ml/src/main/scala/com/linkedin/photon/ml/supervised/classification/SmoothedHingeLossLinearSVMModel.scala:
```diff
@@ -69,3 +70,14 @@ class SmoothedHingeLossLinearSVMModel(override val coefficients: Coefficients)
+   * @return A smoothed hinge loss SVM model
+   */
+  def createModel(coefficients: Coefficients): SmoothedHingeLossLinearSVMModel =
+    new SmoothedHingeLossLinearSVMModel(coefficients)
```
I noticed that this PR also introduces many factory style methods.

Just out of curiosity, what's the motivation behind such choice? Do we plan to hide or protect the implementation details of the constructors?

ashelkovnykov replied:
For the regression model classes (LinearRegressionModel, LogisticRegressionModel, etc.) it's necessary to keep the GeneralizedLinearOptimizationProblem and descendants simple and avoid the type/covariance issues we had prior to the refactor.

For the OptimizerFactory, it's a standard implementation of the Factory Pattern: we don't know what kind of Optimizer to generate until runtime. 

For the optimization problem classes it's just a convenience method, it's not necessary. In these cases the factory methods serve as constructors.

After I noticed that I was using the factory pattern, I decided I may as well restrict object construction for those classes to the factory method. I don't have any hard reasoning, it just seemed natural to fully commit to the factory pattern, also to limit access as strictly as possible and then loosen it if necessary.

XianXing commented on photon-ml/src/main/scala/com/linkedin/photon/ml/function/ObjectiveFunction.scala:
```diff
@@ -0,0 +1,73 @@
+abstract class ObjectiveFunction {
+  type Data
+  type Coefficients
+
```
Minor issue and just out of curiosity: why not use generic type here, for example ```ObjectiveFunction[Data, Coefficients]```?

ashelkovnykov replied:
Generic types were causing issues for the `DiffFunction`, `TwiceDiffFunction`, and `L2Regularization` traits - especially the `L2Regularization` traits due to them being stackable.

XianXing replied:
Hmm, didn't think about this before. Can you given one or two examples for what issues generic types brought? (for my learning purpose and folks who read this PR and are interested).

ashelkovnykov replied:
@XianXing I have not forgotten this question! As part of InDay today, I went through and tried to replace abstract types with generics. I was able to do so - it looks like my previous errors were due to the half-done state of the code in between two simultaneous refactors. However, I still think that abstract types is the way to go. Please take a look at my personal fork - you should be able to see a branch called `xian` which has generics instead.

I feel that the abstract types make the code less cluttered. Yes, in some places the types are implicit instead of explicit, but this is only in locations where the exact types are not important: the relationships between the classes are. The types are kept explicit in the locations where they are bound.

Please let me know what you think, as always I value you input.

XianXing replied:
Thanks for looking into this, it's good to know that the generics are still working in such cases.

I don't have strong opinion on this, as I am also learning best practices about Scala, like the one you explained above.

XianXing commented on photon-ml/src/main/scala/com/linkedin/photon/ml/function/L2Regularization.scala:
```diff
@@ -0,0 +1,175 @@
+   *
+   * @param newRegWeight The new L2 regularization weight
+   */
+  protected[ml] def l2RegularizationWeight_=(newRegWeight: Double): Unit = l2RegWeight = newRegWeight
```
Hmm, nice trick, but some additional documentation might be helpful.

ashelkovnykov replied:
Certainly, will do.

XianXing commented on photon-ml/src/main/scala/com/linkedin/photon/ml/function/L2Regularization.scala:
```diff
@@ -0,0 +1,175 @@
+    input: Data,
+    coefficients: Coefficients,
+    normalizationContext: Broadcast[NormalizationContext]): Double =
+    super.value(input, coefficients, normalizationContext) + l2RegValue(convertToVector(coefficients))
```
Why this function has implementation but still being abstract?

Do we need to make it abstract with the implementations, or do we need an actual implementation for it?

joshvfleming replied:
Yeah. Making this ```abstract``` is necessary for the stacking mechanism to work.

XianXing replied:
Thanks for the explanation. This is interesting, will read the documentation.

XianXing commented on photon-ml/src/main/scala/com/linkedin/photon/ml/function/DistributedObjectiveFunction.scala:
```diff
@@ -0,0 +1,73 @@
+ *                           aggregation. It is experimental currently because treeAggregate is unstable in Spark
+ *                           versions 1.4 and 1.5.
+ */
+abstract class DistributedObjectiveFunction(sparkContext: SparkContext, treeAggregateDepth: Int)
```
I like such design choice- it's long overdue!

XianXing commented on photon-ml/src/main/scala/com/linkedin/photon/ml/function/HessianDiagonalAggregator.scala:
```diff
@@ -0,0 +1,128 @@
+ * @param dim The dimension (number of features) of the aggregator
+ */
+@SerialVersionUID(1L)
+protected[ml] class HessianDiagonalAggregator(func: PointwiseLossFunction, val dim: Int) extends Serializable {
```
As a side note related to #128, the Ads team found some problem with the variance computation, do you have any insights?

TDDFT replied:
Just curious: this doesn't need normalization?

ashelkovnykov replied:
The original code I modified didn't use normalization for the Hessian diagonal. Do you recall why?

The diagonal is only used for computing variances, necessary only at the end of model training and not affecting any subsequent computations. At that time, the optimization problem should have the final, non-normalized coefficient means which can be used directly to compute the variances. Does that sound right?

XianXing commented on photon-ml/src/main/scala/com/linkedin/photon/ml/optimization/DistributedOptimizationProblem.scala:
```diff
@@ -0,0 +1,198 @@
+   * @return The learned generalized linear models of each regularization weight and iteration.
+   */
+  def runWithSampling(input: RDD[(Long, LabeledPoint)], initialModel: GeneralizedLinearModel): GeneralizedLinearModel = {
+    samplerOption match {
```
I think the line length exceeds 120 characters here.

TDDFT commented on photon-ml/src/main/scala/com/linkedin/photon/ml/function/glm/DistributedGLMLossFunction.scala:
```diff
@@ -0,0 +1,168 @@
+ *
+ * with:
+ *
+ * z_i = w^T^ z_i.
```
z_i = w^T^ x_i ?

ashelkovnykov replied:
Good catch, thank you

TDDFT commented on photon-ml/src/main/scala/com/linkedin/photon/ml/function/glm/DistributedGLMLossFunction.scala:
```diff
@@ -0,0 +1,168 @@
+ * [[PointwiseLossFunction]]. Since the loss function could change for different types of normalization, a normalization
+ * context object indicates which normalization strategy is used to evaluate the loss function.
+ *
+ * @param singleLossFunction A single loss function l(z, y) used for the generalized linear model
```
Does singlePointLossFunction sounds better?

TDDFT commented on photon-ml/src/main/scala/com/linkedin/photon/ml/function/glm/IndividualGLMLossFunction.scala:
```diff
@@ -0,0 +1,151 @@
+ *
+ * with:
+ *
+ * z_i = w^T^ z_i.
```
z_i = w^T^ x_i?

TDDFT commented on photon-ml/src/main/scala/com/linkedin/photon/ml/function/glm/IndividualGLMLossFunction.scala:
```diff
@@ -0,0 +1,151 @@
+  override protected[ml] def value(
+    input: Iterable[LabeledPoint],
+    coefficients: Vector[Double],
+    normalizationContext: Broadcast[NormalizationContext]): Double =
```
I guess you must have thought about it thoroughly. But why does this API require a Broadcast object? I didn't quite get it.

ashelkovnykov replied:
There's only one normalization context for the whole training job (for both Photon and GAME, though for GAME it's always `NoNormalization()` as normalization is disabled for GAME currently). To reduce network overhead, it should be broadcast. Even the `IndividualOptimizationProblem` objects need a broadcast normalization context, as though they themselves solve their problem local to one node, they should all use the same normalization context and they are distributed across the cluster.

Note that this may need to be revisited as part of adding normalization to GAME (coming this quarter, I believe).

TDDFT replied:
Cool. Makes sense. Thanks for the explanation.

joshvfleming commented on photon-ml/src/main/scala/com/linkedin/photon/ml/function/svm/DistributedSmoothedHingeLossFunction.scala:
```diff
@@ -0,0 +1,132 @@
+  override protected[ml] def value(
+    input: RDD[LabeledPoint],
+    coefficients: Broadcast[Vector[Double]],
+    normalizationContext: Broadcast[NormalizationContext]): Double =
```
Minor: the indentation here is off: arguments should be indented 4 spaces rather than 2. Check other functions in the PR also, there are a few other cases.

ashelkovnykov replied:
According to [this part](http://docs.scala-lang.org/style/indentation.html#methods-with-numerous-arguments) of the style guide, method arguments are 2 spaces. It's only class members that are 4.

joshvfleming replied:
That's for method calls. Method declarations follow the style of class declarations.

joshvfleming commented on photon-ml/src/main/scala/com/linkedin/photon/ml/function/glm/DistributedGLMLossFunction.scala:
```diff
@@ -0,0 +1,168 @@
+            with L2RegularizationTwiceDiff
+        objectiveFunction.l2RegularizationWeight = regularizationContext
+          .getL2RegularizationWeight(configuration.regularizationWeight)
+        objectiveFunction
```
Slightly more concise:

```scala
new DistributedGLMLossFunction(singleLossFunction, sparkContext, treeAggregateDepth)
  with L2RegularizationTwiceDiff {
  l2RegularizationWeight = regularizationContext
    .getL2RegularizationWeight(configuration.regularizationWeight)
}
```

Same comment for other cases here.

ashelkovnykov replied:
Thank you - that's much nicer

joshvfleming commented on photon-ml/src/main/scala/com/linkedin/photon/ml/function/glm/PointwiseLossFunction.scala:
```diff
@@ -12,17 +12,28 @@
- * of the log-likelihood for one data point, except for an irrelevant constant term.
+ * The function to calculate l(z, y) with:
+ *
+ *    z = theta^T^x + offset
```
Minor: in some places we're using ```w``` for the weights, in others we're using ```theta```. Should be consistent.

joshvfleming commented on photon-ml/src/main/scala/com/linkedin/photon/ml/function/svm/SmoothedHingeLossLinearSVMFunction.scala:
```diff
@@ -12,33 +12,19 @@
-}
-
-object SmoothedHingeLossFunction {
+object SmoothedHingeLossLinearSVMFunction {
```
I'm not sure if renaming this is quite right. A smooth hinge loss function is actually distinct from SVM, though probably its most common application is in SVM classifiers.

ashelkovnykov commented on photon-ml/src/main/scala/com/linkedin/photon/ml/function/DiffFunction.scala:
```diff
@@ -15,307 +15,38 @@
-   */
-  protected[ml] def calculateAt(datum: Datum, coefficients: Vector[Double], cumGradient: Vector[Double]): Double
-
+trait DiffFunction extends ObjectiveFunction {
```
I wrote "mixin" but I meant "stackable" for the L2 regularization traits. It is the DiffFunction and TwiceDiffFunction which are mixin.

XianXing commented on photon-ml/src/main/scala/com/linkedin/photon/ml/ModelTraining.scala:
```diff
@@ -119,37 +119,54 @@ object ModelTraining extends Logging {
+    val (glmConstructor, objectiveFunction) = taskType match {
+      case LOGISTIC_REGRESSION =>
+        val constructor = LogisticRegressionModel.createModel _
+        val objective = DistributedGLMLossFunction.createLossFunction(
```
Sounds good. And maybe we also want to document that ```DistributedGLMLossFunction``` (and other loss functions) can actually create objective function?

TDDFT commented on photon-ml/src/main/scala/com/linkedin/photon/ml/function/ValueAndGradientAggregator.scala:
```diff
@@ -15,24 +15,21 @@
+import com.linkedin.photon.ml.function.glm.PointwiseLossFunction
+import com.linkedin.photon.ml.normalization.NormalizationContext
 
+// TODO: Better document this algorithm, especially normalization.
```
I wrote a wiki page for the algorithm at LinkedIn wiki. @ashelkovnykov do you know what I'm talking about? Since it's open sourced I cannot use the wiki link here. I don't know how to reconcile this.

ashelkovnykov replied:
@TDDFT I know the one. It's very thorough, but like you said I'm not sure how to reconcile it as it's an internal document. I'm going to make it a separate issue for now.

ferd36 commented on photon-ml/src/main/scala/com/linkedin/photon/ml/function/glm/PoissonLossFunction.scala:
```diff
@@ -51,12 +44,10 @@ object PointwisePoissonLossFunction extends PointwiseLossFunction {
    * @param label The label, i.e. y in l(z, y)
    * @return The value and the 2st derivative with respect to z
    */
-  override def d2lossdz2(margin: Double, label: Double): Double = {
```
Dzz? IMHO might be more readable? https://www.wikiwand.com/en/Notation_for_differentiation#/Euler.27s_notation

ashelkovnykov replied:
To be clear, you'd like to rename `d2lossdz2` to `DzzLoss`, is that correct?

ashelkovnykov replied:
@ferd36 As per our offline talk, I'll make this change, and log an issue related to refactoring the relationships between `TRON`, `TwiceDiffFunction`, and `PointwiseLossFunction` (and derived classes).

ferd36 commented on photon-ml/src/main/scala/com/linkedin/photon/ml/function/svm/SmoothedHingeLossLinearSVMFunction.scala:
```diff
@@ -12,33 +12,19 @@
  * under the License.
  */
-package com.linkedin.photon.ml.function
+package com.linkedin.photon.ml.function.svm
```
I'm surprised that package "svm" is inside package "function". I thought "svm" referred to "support vector machine", which is a whole model. Wikipedia for example says: "In machine learning, support vector machines are **supervised learning models** with associated learning algorithms that analyze data used for classification and regression analysis." So, they would not be "functions"?

ashelkovnykov replied:
This class is a bit wiggy, as it's a function that we use with our GLM framework to approximate an SVM. What would you recommend?

ferd36 commented on photon-ml/src/main/scala/com/linkedin/photon/ml/optimization/IndividualOptimizationProblem.scala:
```diff
@@ -0,0 +1,139 @@
+  }
+}
+
+object IndividualOptimizationProblem {
```
That "Individual" again :-)

ashelkovnykov commented on photon-ml/src/main/scala/com/linkedin/photon/ml/ModelTraining.scala:
```diff
@@ -119,37 +119,54 @@ object ModelTraining extends Logging {
+
+      case LINEAR_REGRESSION =>
+        val constructor = LinearRegressionModel.createModel _
+        val objective = DistributedGLMLossFunction.createLossFunction(
```
I'm okay with these changes, and if no one objects will implement them.

ashelkovnykov commented on photon-ml/src/main/scala/com/linkedin/photon/ml/cli/game/training/Driver.scala:
```diff
@@ -183,9 +189,11 @@ final class Driver(val params: Params, val sparkContext: SparkContext, val logge
    * @return The training evaluator
    */
   protected[training] def prepareTrainingLossFunctionEvaluator(gameDataSet: RDD[(Long, GameDatum)]): Evaluator = {
```
I'll leave this alone for now since it's being handled in your PR

ashelkovnykov commented on photon-ml/src/main/scala/com/linkedin/photon/ml/cli/game/training/Driver.scala:
```diff
@@ -512,12 +643,23 @@ final class Driver(val params: Params, val sparkContext: SparkContext, val logge
       timer.start()
-      val validatingDataAndEvaluatorOption = validatingDataAndEvaluatorsOption
-          .map { case (validationData, evaluators) => (validationData, evaluators.head) }
+      val validatingDataAndEvaluatorOption =
```
I'll leave this alone for now since it's being handle in your PR

ashelkovnykov commented on photon-ml/src/main/scala/com/linkedin/photon/ml/algorithm/FactoredRandomEffectCoordinate.scala:
```diff
@@ -53,23 +76,35 @@ protected[ml] class FactoredRandomEffectCoordinate[GLM <: GeneralizedLinearModel
+    model: DatumScoringModel): (DatumScoringModel, Option[OptimizationTracker]) = model match {
+      case factoredRandomEffectModel: FactoredRandomEffectModel =>
+        val sparkContext = randomEffectDataSet.sparkContext
+        val randomEffectOptimizationProblem = factoredRandomEffectOptimizationProblem.randomEffectOptimizationProblem
```
I've modified this one a little, but in my opinion there's not too much room for change. If you've got a suggestion, I'm open to it

ashelkovnykov commented on photon-ml/src/main/scala/com/linkedin/photon/ml/algorithm/RandomEffectCoordinate.scala:
```diff
@@ -14,85 +14,131 @@
 /**
  * The optimization problem coordinate for a random effect model
  *
+ * @tparam Function The type of objective function used to solve individual random effect optimization problems
```
I went with `Objective`, it's more specific but still short (as per our offline conversation)

ashelkovnykov commented:
I have just updated this PR with performance comparison information. The performance comparison was completed some time ago, but needed to be redone (due to a performance bug which was discovered and addressed) and posted.

XianXing commented:
This is exciting, ~37% speed improvement is huge.

When you have time, would you highlight a few things that may lead to such lift in this PR for documentation purpose, and what is the performance bug we had before?
